File: packages/argentum-node/src/sse.ts
/**
 * A server-sent events reader.
 *
 * Written rather than depended on, for the reason the whole package has no
 * runtime dependencies: SSE is a line-oriented format with four field names,
 * and a client that pulls in a parser for it has taken on a supply chain to
 * avoid sixty lines.
 *
 * What it does that a naive split on `\n\n` does not:
 *
 * - reassembles frames across chunk boundaries, because a network read has no
 *   relationship to a frame boundary;
 * - keeps the last `id:` it saw, which is what a caller sends back as
 *   `Last-Event-ID` to resume;
 * - drops comments (`: heartbeat`) silently, which is what makes the keepalive
 *   cost an integrator no code.
 */
export interface SSEFrame {
  event: string;
  data: string;
  id?: string;
}

/** Parses one HTTP response body into frames, in order. */
export async function* readSSE(res: Response, signal?: AbortSignal): AsyncGenerator<SSEFrame> {
  if (!res.body) return;
  const decoder = new TextDecoder();
  let buffer = '';

  // @ts-expect-error — Node's ReadableStream is async-iterable at runtime from
  // v18, but the DOM lib this package compiles against does not declare it.
  for await (const chunk of res.body) {
    if (signal?.aborted) return;
    buffer += decoder.decode(chunk as Uint8Array, { stream: true });

    let split: number;
    // `\n\n` only: this API writes `\n` line endings, and accepting `\r\n\r\n`
    // as well would mean handling the mixed case, which nothing produces.
    while ((split = buffer.indexOf('\n\n')) !== -1) {
      const raw = buffer.slice(0, split);
      buffer = buffer.slice(split + 2);
      const frame = parseFrame(raw);
      if (frame) yield frame;
    }
  }
}

function parseFrame(raw: string): SSEFrame | undefined {
  let event = 'message';
  let id: string | undefined;
  const data: string[] = [];

  for (const line of raw.split('\n')) {
    if (line === '' || line.startsWith(':')) continue; // a comment, i.e. the heartbeat
    const colon = line.indexOf(':');
    const field = colon === -1 ? line : line.slice(0, colon);
    // One optional space after the colon is part of the format, not part of
    // the value — a JSON payload that lost its first character would fail to
    // parse for a reason nobody would find quickly.
    const value = colon === -1 ? '' : line.slice(colon + 1).replace(/^ /, '');
    switch (field) {
      case 'event':
        event = value;
        break;
      case 'data':
        data.push(value);
        break;
      case 'id':
        id = value;
        break;
      default:
        break; // `retry:` and anything we do not know
    }
  }

  if (data.length === 0 && event === 'message') return undefined;
  return { event, data: data.join('\n'), ...(id === undefined ? {} : { id }) };
}

File: packages/argentum-python/src/argentum/_sse.py
"""A server-sent events reader, shared by both clients.

Written rather than depended on: SSE is a line-oriented format with four field
names, and pulling in a parser for it would double this package's dependency
count.

What it does that a naive ``split`` does not: it drops comments (``:
heartbeat``) silently, so the keepalive costs an integrator no code, and it
keeps the ``id:`` — which is what a caller sends back as ``Last-Event-ID`` to
resume a dropped stream.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Dict, Iterable, List, Optional


@dataclass
class Event:
    """One SSE frame, with its payload already decoded.

    ``event`` is the discriminator: ``delta``, ``final``, ``progress``,
    ``report``, and so on. Treat an unknown one as ignorable — that is what lets
    the server add frames without breaking you.
    """

    event: str
    data: Dict[str, Any] = field(default_factory=dict)
    #: Present only on frames backed by something persisted — a `message` or a
    #: `final`. A delta has none, because it exists nowhere but the connection
    #: that carried it.
    id: Optional[str] = None

    @property
    def terminal(self) -> bool:
        """True for the frames after which the server closes the stream."""
        return self.event in ("final", "error", "report")


class _FrameAssembler:
    """Turns a sequence of lines into frames. Fed by whichever client is reading."""

    def __init__(self) -> None:
        self._lines: List[str] = []

    def feed(self, line: str) -> Optional[Event]:
        # httpx strips the newline, so a blank string is the frame boundary.
        if line != "":
            self._lines.append(line)
            return None
        raw, self._lines = self._lines, []
        return _parse(raw)


def _parse(lines: Iterable[str]) -> Optional[Event]:
    name = "message"
    ident: Optional[str] = None
    data: List[str] = []

    for line in lines:
        if not line or line.startswith(":"):
            continue  # a comment, i.e. the heartbeat
        field_name, _, value = line.partition(":")
        # One optional space after the colon is part of the format, not part of
        # the value — a JSON payload that lost its first character would fail to
        # parse for a reason nobody would find quickly.
        value = value[1:] if value.startswith(" ") else value
        if field_name == "event":
            name = value
        elif field_name == "data":
            data.append(value)
        elif field_name == "id":
            ident = value

    if not data:
        return None
    payload = "\n".join(data)
    try:
        decoded = json.loads(payload)
    except json.JSONDecodeError:
        decoded = {"raw": payload}
    return Event(event=name, data=decoded if isinstance(decoded, dict) else {"value": decoded}, id=ident)
